src/Api/Serializer/RestSerializer.php

```diff
@@ -40,17 +44,47 @@ public function __construct(Service $api, $endpoint)
      *
      * @return RequestInterface
      */
-    public function __invoke(CommandInterface $command)
+    public function __invoke(
+        CommandInterface $command,
+        $endpointProvider = null,
+        array $clientArgs = null
+    )
     {
-        $operation = $this->api->getOperation($command->getName());
-        $args = $command->toArray();
-        $opts = $this->serialize($operation, $args);
-        $uri = $this->buildEndpoint($operation, $args, $opts);
+        $operationName = $command->getName();
+        $operation = $this->api->getOperation($operationName);
+        $commandArgs = $command->toArray();
+        $opts = $this->serialize($operation, $commandArgs);
+        $headers = isset($opts['headers']) ? $opts['headers'] : [];
+
+        if ($endpointProvider instanceof EndpointProvider) {
+            $service = $this->api->getServiceName();
+            $providerArgs = $this->resolveProviderArgs(
+                $operation,
+                $endpointProvider,
+                $commandArgs,
+                $clientArgs,
+                $operationName
+            );
+            $endpoint = $endpointProvider->resolveEndpoint($providerArgs);
+            $this->endpoint = new Uri($endpoint->getUrl());
+
+            if ($service === 's3') {
```

Comment:
Unfortunately S3 can't remove `Bucket` from the path in their models for backward-compatibility concerns.  We either have to pre-process the model to remove it, or ignore it like we're doing here.  For S3control, we have to remove `AccountId` from the URI and also remove the `required` trait from it, but that's done at client instantiation

Reply:
I'm on board with an s3-specific customization instead of doing it here.  I've already done pre-processing with the S3control model, so I could just do that on the S3 client at instantiation if we're using the new provider